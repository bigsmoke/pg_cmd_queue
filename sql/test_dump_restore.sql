\o /dev/null
select  not :{?test_stage} as test_stage_missing
        ,not :{?extension_name} as extension_name_missing;
\o
\gset
\if :test_stage_missing
    \warn 'Missing `:test_stage` variable.'
    \quit
\endif
\if :extension_name_missing
    \warn 'Missing `:extension_name` variable.'
    \quit
\endif
\o /dev/null
select  :'test_stage' = 'pre-dump' as in_pre_dump_stage
        ,:'test_stage' = 'pre-restore' as in_pre_restore_stage;
\o
\gset

\set SHOW_CONTEXT 'errors'

\if :in_pre_restore_stage
    -- Let's generate some noise to offset the OIDs, to ensure that we're not relying on OIDs remaining the
    -- same between the moment of `pg_dump` and the moment of `pg_restore`.
    do $$
    declare
        _i int;
    begin
        for
            _i
        in select
            s.i
        from
            generate_series(1, 10) as s(i)
        loop
            execute format('CREATE TABLE test_dump_restore__oid_noise__tbl_%s (a int)', _i);
            execute format('CREATE TYPE test_dump_restore__oid_noise__rec_%s AS (a int, b int)', _i);
            execute format(
                'CREATE FUNCTION test_dump_restore__oid_noise__func_%s() RETURNS int RETURN 1'
                ,_i
            );
        end loop;
    end;
    $$;

    \quit
\endif

-- We put the `psql` variables into a temporary table, so that we can read them out from within the
-- PL/pgSQL`DO` block, as we cannot access these variables from within PL/pgSQL.
select
    :'extension_name' as extension_name
    ,:'test_stage' as test_stage
into temporary
    vars
;

do $$
declare
    _extension_name text := (select extension_name from vars);
    _test_stage text := (select test_stage from vars);
    _test_proc text;
begin
    if _test_stage = 'pre-dump' then
        execute format('CREATE EXTENSION %I WITH CASCADE', _extension_name);
    end if;

    for
        _test_proc
    in
    select
        case
            when pg_proc.prokind = 'p' then
                'CALL ' || pg_proc.oid::regproc::text
            else
                'PERFORM ' || pg_proc.oid::regproc::text
        end || '(' || quote_literal(_test_stage) || ')'
    from
        pg_depend
    inner join
        pg_proc
        on pg_proc.oid = pg_depend.objid
        and pg_depend.classid = 'pg_proc'::regclass
    where
        pg_depend.refclassid = 'pg_extension'::regclass
        and pg_depend.refobjid = (select oid from pg_extension where extname = _extension_name)
        and pg_proc.proname like 'test\_dump\_restore\_\_%'
        and pg_proc.prokind in ('f', 'p')
    loop
        raise notice '%', _test_proc;
        execute _test_proc;
    end loop;
end;
$$;
