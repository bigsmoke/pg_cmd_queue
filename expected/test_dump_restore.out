-- createdb
-- psql -X -f 'sql/test_dump_restore.sql' -v 'extension_name=pg_cmd_queue' -v 'test_stage=pre-dump'
SELECT 1
psql:sql/test_dump_restore.sql:94: NOTICE:  installing required extension "hstore"
psql:sql/test_dump_restore.sql:94: NOTICE:  CALL cmdq.test_dump_restore__pg_cmd_queue('pre-dump')
DO
-- pg_dumpall --roles-only
-- pg_dump --format=custom --file <dump_file>
-- kill -9 <pg_pid>
-- rm -r $PGDATA
-- psql postgres -X -c '\set ON_ERROR_STOP' -f <roles_dump_file>
SET
SET
SET
ALTER ROLE
-- createdb 'test_dump_restore_oid_noise'
-- psql -X -f 'sql/test_dump_restore.sql' -v 'extension_name=pg_cmd_queue' -v 'test_stage=pre-restore'
DO
-- pg_restore --create --dbname postgres <dump_file>
-- psql -X -f 'sql/test_dump_restore.sql' -v 'extension_name=pg_cmd_queue' -v 'test_stage=post-restore'
SELECT 1
psql:sql/test_dump_restore.sql:94: NOTICE:  CALL cmdq.test_dump_restore__pg_cmd_queue('post-restore')
DO
